### PR TITLE
fix: Retrieve authenticated manifest packages 

### LINKED ISSUE #402: Access private Git repositories via manifest
### Feature request summary

This issue tracks accessing private Git repositories via the manifest package manager. This in turn will allow retrieving source code via `kraft pkg pull` and `kraft build`.

### Describe alternatives

_No response_

### Related architectures

None

### Related platforms

None

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nderjung**
### Prerequisite checklist

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

This PR adds relevant functionality towards retrieving components from remote sources which require authentication.  It is possible to use HTTP Basic and Bearer to access archives from GitHub as well as Git over SSH.

To test this PR, set the source of the unikraft core (or library) to a private Git repository:

```yaml
---
specification: '0.5'
name: helloworld
unikraft:
  source: ssh://github.com/username/unikraft.git
targets:
  - architecture: x86_64
    platform: kvm
```

Then in a terminal:

```
eval `ssh-agent` # Initialize a new SSH agent socket
ssh-add path/to/ssh_key # Add a private key to the agent
cd path/to/app 
kraft pkg pull # Pull the private components over SSH
```

To test with GitHub authorization header, login to GitHub with a personal access token:

```
kraft login -u $USERNAME -t $TOKEN github.com
```

Then the path can then just be a normal HTTPS schema, e.g.:

```yaml
---
specification: '0.5'
name: helloworld
unikraft:
  source: https://github.com/username/unikraft.git
targets:
  - architecture: x86_64
    platform: kvm
```

GitHub-Closes: #402

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -138,17 +139,36 @@ properclean: ## Completely clean the repository's build artifacts.
...
-
-$(VENDORDIR)/github.com/libgit2/git2go/v31/static-build/install/lib/pkgconfig/libgit2.pc: $(VENDORDIR)/github.com/libgit2/git2go/v31/vendor/libgit2
-	$(MAKE) -C $(VENDORDIR)/github.com/libgit2/git2go/v31 install-static
+GIT2GO_VERSION ?= v31
 
-$(VENDORDIR)/github.com/libgit2/git2go/v31/vendor/libgit2: $(VENDORDIR)/github.com/libgit2/git2go
-	$(GIT) -C $(VENDORDIR)/github.com/libgit2/git2go/v31 submodule update --init --recursive
+.PHONY: git2go
+git2go: $(VENDORDIR)/github.com/libgit2/git2go/$(GIT2GO_VERSION)/static-build/install/lib/pkgconfig/libgit2.pc
+	(cd $(VENDORDIR)/github.com/libgit2/git2go/$(GIT2GO_VERSION) && $(GO) install --tags "static" ./...)
```
Comment (antoineco) [inline Makefile]: This works for me locally:

```suggestion
	$(GO) install -tags static github.com/libgit2/git2go/v31/...
```

You can pass the `-x` flag to see all the executed commands and verify. Sample output with `-x`:

```
# get https://proxy.golang.org/golang.org/x/crypto/@v/v0.6.0.info: 200 OK (0.014s)
WORK=/tmp/go-build3675054337
mkdir -p $WORK/b005/
cat >/tmp/go-build3675054337/b005/importcfg << 'EOF' # internal
# import config
EOF
cd /home/acotten/git/unikraft/kraftkit
/home/acotten/.local/share/go/pkg/tool/linux_amd64/compile -o $WORK/b005/_pkg_.a -trimpath "$WORK/b005=>" -p internal/goarch -std -+ -complete -buildid fcyczH0YXp2IMM4KyVvj/fcyczH0YXp2IMM4KyVvj -goversion go1.20.3 -c=4 -nolocalimports -importcfg $WORK/b005/importcfg -pack /home/acotten/.local/share/go/src/internal/goarch/goarch.go /home/acotten/.local/share/go/src/internal/goarch/goarch_amd64.go /home/acotten/.local/share/go/src/internal/goarch/zgoarch_amd64.go
mkdir -p $WORK/b012/


...
cd /home/acotten/git/unikraft/kraftkit/third_party/github.com/libgit2/git2go/v31
/home/acotten/.local/share/go/pkg/tool/linux_amd64/pack r $WORK/b001/_pkg_.a $WORK/b001/_x001.o $WORK/b001/_x002.o $WORK/b001/_x003.o $WORK/b001/_x004.o $WORK/b001/_x005.o $WORK/b001/_x006.o $WORK/b001/_x007.o $WORK/b001/_x008.o $WORK/b001/_x009.o $WORK/b001/_x010.o $WORK/b001/_x011.o $WORK/b001/_x012.o $WORK/b001/_x013.o $WORK/b001/_x014.o $WORK/b001/_x015.o $WORK/b001/_x016.o $WORK/b001/_x017.o $WORK/b001/_x018.o $WORK/b001/_x019.o $WORK/b001/_x020.o $WORK/b001/_x021.o $WORK/b001/_x022.o $WORK/b001/_x023.o $WORK/b001/_x024.o $WORK/b001/_x025.o $WORK/b001/_x026.o $WORK/b001/_x027.o $WORK/b001/_x028.o $WORK/b001/_x029.o $WORK/b001/_x030.o $WORK/b001/_x031.o $WORK/b001/_x032.o $WORK/b001/_x033.o $WORK/b001/_x034.o $WORK/b001/_x035.o $WORK/b001/_x036.o $WORK/b001/_x037.o $WORK/b001/_x038.o $WORK/b001/_x039.o $WORK/b001/_x040.o $WORK/b001/_x041.o $WORK/b001/_x042.o $WORK/b001/_x043.o $WORK/b001/_x044.o $WORK/b001/_x045.o $WORK/b001/_x046.o $WORK/b001/_x047.o $WORK/b001/_x048.o $WORK/b001/_x049.o # internal
/home/acotten/.local/share/go/pkg/tool/linux_amd64/buildid -w $WORK/b001/_pkg_.a # internal
cp $WORK/b001/_pkg_.a /home/acotten/.cache/go-build/11/11f08447963fcea23b664273ff35500e6a9b5474639dacf27ab3ccfc5f25bca6-d # internal
```
```

File: Makefile [hunk 2]
```diff
@@ -138,17 +139,36 @@ properclean: ## Completely clean the repository's build artifacts.
...
+		-DDEPRECATE_HARD="${BUILD_DEPRECATE_HARD}" \
+		$(VENDORDIR)/github.com/libgit2/git2go/$(GIT2GO_VERSION)/vendor/libgit2)
+	$(MAKE) -C $(VENDORDIR)/github.com/libgit2/git2go/$(GIT2GO_VERSION)/static-build/build install
+
+$(VENDORDIR)/github.com/libgit2/git2go/$(GIT2GO_VERSION)/vendor/libgit2: $(VENDORDIR)/github.com/libgit2/git2go
+	$(GIT) -C $(VENDORDIR)/github.com/libgit2/git2go/$(GIT2GO_VERSION) submodule update --init --recursive
 
 $(VENDORDIR)/github.com/libgit2/git2go:
-	$(GIT) clone --branch v31.7.9 --recurse-submodules https://github.com/libgit2/git2go.git $@/v31
+	$(GIT) clone --branch $(GIT2GO_VERSION).7.9 --recurse-submodules https://github.com/libgit2/git2go.git $@/$(GIT2GO_VERSION)
```
Comment (antoineco) [inline Makefile]: This `GIT2GO_VERSION` feels error prone and very unnecessary to me. For example, there is no guarantee that `v32.7.9` exists here (considering `v31 -> v32`), so the update won't be as smooth as bumping this value anyway.

I suggest that we remove this part of the path entirely and vendor the lib in `third_party/libgit2/git2go`.

1. The `replace` in `go.mod` will make it work regardless.
2. The version will be obvious from the `go.mod` file.

Keep it simple, this module is already complex enough.

File: Makefile (hunk 1 above)
Comment (nderjung) [reply Makefile]: Good catch! I tried a few variants but this one does the trick! Will incorporate in a fixup. :)

File: Makefile (hunk 2 above)
Comment (nderjung) [reply Makefile]: Makes sense, you're right.

I'll simplify the gitmodule in a separate commit and rebase appropriately with your suggested changes.

File: manifest/pack_pull_archive.go [hunk 3]
```diff
@@ -89,13 +120,23 @@ func pullArchive(ctx context.Context, manifest *Manifest, opts ...pack.PullOptio
...
+
+		head.Header.Set("User-Agent", version.UserAgent())
+		if len(authHeader) > 0 {
+			head.Header.Set("Authorization", authHeader)
+		}
+
 		log.G(ctx).WithFields(logrus.Fields{
 			"url":    resource,
-			"method": "HEAD",
+			"method": "get",
```
Comment (craciunoiuc) [inline manifest/pack_pull_archive.go]:
```suggestion
			"method": "GET",
```

**Review by craciunoiuc — COMMENTED** [review]
One small thing from my side

File: .gitmodules [hunk 4]
```diff
@@ -1,3 +1,3 @@
 [submodule "libgit2/git2go"]
-	path = third_party/github.com/libgit2/git2go/v31
+	path = third_party/libgit2/git2go
```
Comment (antoineco) [inline .gitmodules:2]: I love to see that <3

File: manifest/pack_pull_archive.go (hunk 3 above)
Comment (antoineco) [reply manifest/pack_pull_archive.go]: nit: Even better `http.MethodGet`.

**Review by antoineco — APPROVED** [review]
Code LGTM.
I haven't looked how to test that manually, but all changes are thoroughly tested by their author, right.... right?

Approved-by: Antoine Cotten <antoine@unikraft.io>

File: .gitmodules (hunk 4 above)
Comment (nderjung) [reply .gitmodules:2]: 🫶

**Comment by craciunoiuc** [comment]
He is not to be trusted 👀

**Review by craciunoiuc — APPROVED** [review]
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

**Pull request merged**